### ISSUE TITLE
Potential performance hit from PipeStream misuse

### ISSUE THREAD

**Issue by cdmihai** [Microsoft/msbuild#884]
MSBuild is using a [NamedPipeServerStream](https://github.com/Microsoft/msbuild/blob/master/src/Shared/NodeEndpointOutOfProcBase.cs#L486) wrapped by a [binary reader](https://github.com/Microsoft/msbuild/blob/master/src/Shared/InterningBinaryReader.cs) for inter process object marshalling during multiproc builds.

VS perf tests pointed out that deserialization is taking a long time due to a potential misuse of the pipe stream (or it's interaction with the custom interning binary reader): _"the perf trace shows that the code spent 2/3 of time to call the API to read to a buffer the size of a single byte, and 1/3 of time to read a string.  Behind it, the pipeStream code calls async API, and wait it to be processed in another thread for every those bytes."_

```
Name                                                	Inc %	     Inc	 Inc Ct	Exc %	Exc	Exc Ct	Fold	Fold Ct	                            When	    First	      Last
system.core.ni!PipeStream.ReadCore                 	 5.7	28,580.086	415,755	 0.0	 0	    0	  0	     0	_50999999999999999999999999993__	1,467.629	30,203.703
+ system.core.ni!PipeStream.EndRead                 	 5.5	27,873.910	404,585	 0.0	 0	    0	  0	     0	_29999999999999999999999999993__	1,467.629	30,203.703
|+ mscorlib.ni!System.Threading.WaitHandle.WaitOne()	 5.5	27,687.170	402,319	 0.0	 0	    0	  0	     0	_29999999999999999999999999993__	1,467.629	30,203.703

Thread Time caller:
Name                                                                  	Inc %	     Inc	 Inc Ct	Exc %	Exc	Exc Ct	Fold	Fold Ct	                            When	    First	      Last
system.core.ni!PipeStream.ReadCore                                   	 5.7	28,580.086	415,755	 0.0	 0	    0	  0	     0	_50999999999999999999999999993__	1,467.629	30,203.703
+ system.core.ni!System.IO.Pipes.PipeStream.ReadByte()                	 3.9	19,753.338	292,850	 0.0	 0	    0	  0	     0	_26666666666666666666666666662__	1,467.629	30,203.703
+ system.core.ni!PipeStream.Read                                      	 1.8	8,826.576	122,902	 0.0	 0	    0	  0	     0	_02333333333333323323333323331__	1,490.325	30,203.515
|+ microsoft.build.ni!InterningBinaryReader.ReadString                	 1.5	7,764.039	107,894	 0.0	 0	    0	  0	     0	_02222222222222222222222222221__	1,493.011	30,203.515

CPU stack (indicates called many time, not blocked in one call)
Name                                     	Inc %	     Inc	Exc %	Exc	Fold	                            When	    First	      Last
system.core.ni!PipeStream.ReadCore      	53.9	  17,090	 0.1	36	  0	_2066656656566666655555666665554	1,489.024	30,203.740
+ system.core.ni!PipeStream.EndRead      	24.9	   7,902	 0.3	92	  0	_0222222232222222222222222222222	1,489.024	30,201.739
+ system.core.ni!PipeStream.BeginReadCore	19.0	   6,043	 0.2	60	  0	_0222222212222222221222222222111	1,492.010	30,194.837

```

Need to investigate how / why this happens and fix it.

**Comment by cdmihai** [comment]
@dsplaisted I vaguely remember you also spotted some weird pipe stream usage while turning `/m` on on xplat. Does this sound familiar?

**Comment by rainersigwald** [comment]
This should have been resolved by @dsplaisted in #1227.
